ydb/core/tx/datashard/sample_k.cpp

```diff
@@ -0,0 +1,331 @@
+
+private:
+    STFUNC(StateWork) {
+        switch (ev->GetTypeRewrite()) {
+            default:
+                LOG_ERROR(*TlsActivationContext, NKikimrServices::TX_DATASHARD,
+                          "TSampleKScan: StateWork unexpected event type: %" PRIx32 " event: %s",
+                          ev->GetTypeRewrite(), ev->ToString().data());
+        }
+    }
+
+    void SetRow(const TRow& row, ui64 p) {
+        std::pop_heap(MaxRows.begin(), MaxRows.end());
+        DataRows[MaxRows.back().i] = TSerializedCellVec::Serialize(*row);
+        MaxRows.back().p = p;
+        std::push_heap(MaxRows.begin(), MaxRows.end());
+        MaxProbability = MaxRows.front().p;
+    }
+
+    void SendResponse() {
+        if (!Response) {
+            return;
+        }
+        auto ctx = TActivationContext::AsActorContext().MakeFor(SelfId());
+        std::sort(MaxRows.begin(), MaxRows.end());
+        for (auto& [p, i] : MaxRows) {
+            Response->Record.AddProbabilities(p);
+            Response->Record.AddRows(std::move(DataRows[i]));
+        }
+        Response->Record.SetStatus(NKikimrTxDataShard::TEvSampleKResponse::DONE);
+        // TODO(mbkkt) some retry? how?
+        ctx.Send(ResponseActorId, Response.Release());
+    }
+};
+
+class TDataShard::TTxHandleSafeSampleKScan: public NTabletFlatExecutor::TTransactionBase<TDataShard> {
+public:
+    TTxHandleSafeSampleKScan(TDataShard* self, TEvDataShard::TEvSampleKRequest::TPtr&& ev)
+        : TTransactionBase(self)
+        , Ev(std::move(ev))
+    {
+    }
+
+    bool Execute(TTransactionContext&, const TActorContext& ctx) {
+        Self->HandleSafe(Ev, ctx);
+        return true;
+    }
+
+    void Complete(const TActorContext&) {
+        // nothing
+    }
+
+private:
+    TEvDataShard::TEvSampleKRequest::TPtr Ev;
+};
+
+void TDataShard::Handle(TEvDataShard::TEvSampleKRequest::TPtr& ev, const TActorContext&) {
+    Execute(new TTxHandleSafeSampleKScan(this, std::move(ev)));
+}
+
+void TDataShard::HandleSafe(TEvDataShard::TEvSampleKRequest::TPtr& ev, const TActorContext& ctx) {
+    const auto& record = ev->Get()->Record;
+    TRowVersion rowVersion(record.GetSnapshotStep(), record.GetSnapshotTxId());
+
+    // Note: it's very unlikely that we have volatile txs before this snapshot
+    if (VolatileTxManager.HasVolatileTxsAtSnapshot(rowVersion)) {
+        VolatileTxManager.AttachWaitingSnapshotEvent(rowVersion,
+                                                     std::unique_ptr<IEventHandle>(ev.Release()));
+        return;
+    }
+    const ui64 id = record.GetId();
+
+    auto response = MakeHolder<TEvDataShard::TEvSampleKResponse>();
+    response->Record.SetId(id);
+    response->Record.SetTabletId(TabletID());
+
+    TScanRecord::TSeqNo seqNo = {record.GetSeqNoGeneration(), record.GetSeqNoRound()};
+    response->Record.SetRequestSeqNoGeneration(seqNo.Generation);
+    response->Record.SetRequestSeqNoRound(seqNo.Round);
+
+    auto badRequest = [&](const TString& error) {
```

Comment:
badRequest is called many times. Please add some unit tests.

Reply:
It's copied from build index, but yeah, ok

Reply:
Added